server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,278 @@
+    private ArrayList<DocInfo> cachedDocs;
+    private int avgBucketDistance;
+
+    private List<InternalVariableWidthHistogram.Bucket> buckets;
+
+    VariableWidthHistogramAggregator(String name, AggregatorFactories factories, int numBuckets, int shardSize,
+                                     int cacheLimit, @Nullable ValuesSource.Numeric valuesSource,
+                                     DocValueFormat formatter, SearchContext context, Aggregator parent,
+                                     List<PipelineAggregator> pipelineAggregators,
+                                     Map<String, Object> metaData) throws IOException{
+        super(name, factories, context, parent, pipelineAggregators, metaData);
+
+        this.numBuckets = numBuckets;
+        this.valuesSource = valuesSource;
+        this.formatter = formatter;
+
+        this.shardSize = shardSize;
+        this.cacheLimit = cacheLimit;
+        this.cachedDocs = new ArrayList<>();
+        this.buckets = new ArrayList<>();
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        if (valuesSource != null && valuesSource.needsScores()) {
+            return ScoreMode.COMPLETE;
+        }
+        return super.scoreMode();
+    }
+
+    /**
+     * Represents the information needed to collect a document into a bucket.
+     * This class enables us to store a small number of documents in memory before bucketing them
+     */
+    private class DocInfo implements Comparable<DocInfo>{
+
+        int doc;
+        double val;
+        LeafBucketCollector sub;
+
+        DocInfo(int doc, double val, LeafBucketCollector sub){
+            this.doc = doc;
+            this.val = val;
+            this.sub = sub;
+        }
+
+        @Override
+        public int compareTo(DocInfo o) {
+            return Double.compare(val, o.val);
+        }
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        if (valuesSource == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        final SortedNumericDoubleValues values = valuesSource.doubleValues(ctx);
+        return new LeafBucketCollectorBase(sub, values){
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                assert bucket == 0;
+                if(values.advanceExact(doc)){
+                    final int valuesCount = values.docValueCount();
+                    double prevVal = Double.NEGATIVE_INFINITY;
+                    for (int i = 0; i < valuesCount; ++i) {
+                        double val = values.nextValue();
+                        assert val >= prevVal;
+                        if (val == prevVal){
+                            continue;
+                        }
+                        DocInfo docInfo = new DocInfo(doc, val, sub);
```

Comment:
I think we're going to run into troubles with this `DocInfo` class / caching.  Mainly because there are some subtle behaviors which are really hard to notice from the outside (quirks of how Lucene works basically).

Two issues come to mind:

1. `doc` resets to zero for each segment (e.g. each time `getLeafCollector()` is called).  So if a bunch of small segments are collected, but we are still under the cache limit, the cached ArrayList will hold a bunch of docs that all share the same ID
2. The second issue is that regular (non-deferring) aggs execute in "depth-first" mode.  A document is aggregated into a bucket, then sub-aggs are called on the document, then sub-aggs to that, etc etc.  It traverses to the bottom of the agg tree before surfacing and starting over with the next document.  The problem is that once we leave a segment, we can no longer collect that document.  Lucene's data access model is an iterator, which only advances forward and is specific to a particular segment.  Once a different segment is loaded, the old iterator (and associated document data) is gone.  This is unfortunately not very obvious from the leaf collector API.

    So if we cache a value, leave the segment, the underlying doc value iterators are no longer available because we're in a new segment.  Which will prevent us from collecting sub-aggs at a later point.  If you were to call `collectBucket()` on cached documents, the sub-aggs would throw an exception due to iterators not lining up with documents, wrong values being returned, etc.

Instead of extending `BucketsAggregator`, we probably will need to extend `DeferableBucketAggregator`.  This is a type of agg which specifies a "deferring" collector (probably `MergingBucketsDeferringCollector` in this case), which "caches" values/buckets so that we can execute "breadth-first".  Instead of traversing to the bottom of the agg tree, deferring collectors allow one aggregation to fully process its layer of the tree, prune the buckets, then only recurse downwards with the surviving buckets (it's a memory-saving strategy for aggs that generate many buckets).  

You [can see here](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollector.java#L126-L127) that it basically saves the doc ID (as a delta from the last doc) and the bucket, and when the [segment finishes we store it in a list](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollector.java#L103).  Later, when we need to run the deferred sub-aggs, we go back and [re-load the segment context and replay the docs](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollector.java#L170-L204) through a new set of LeafCollectors.  That's how we get around losing the iterators after a segment... we go back and reload the iterators for the sub-aggs.

Getting back to the PR, I think what we should do is defer collection (via `DeferableBucketAggregator`) and collect the first `cacheLimit` values to unique bucket ordinals (`0, 1, 2, ... cacheLimit)`.  That way each doc is saved distinctly and we can re-shuffle them later.  Concurrently we need to save the value, probably to a `BigDoubleArray` (`bigArrays.newDoubleArray(...)`) which tracks the "centroid".  First document corresponds to first entry, etc.    

Then once we hit `cacheLimit`, we stop, figure out the real centroids, construct a merge map and call [`MergingBucketsDeferringCollector#mergeBuckets()`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/MergingBucketsDeferringCollector.java#L55).  This will do the right magic in the background to collect our `[0,1,2,.. cacheLimit]` bucket ordinals into useful buckets.  E.g. the merge map might look like `[1,1,5,2,2,5,1]`, meaning the first/second/seventh existing bucket ordinals are re-mapped to ordinal 1, the third bucket ordinal is remapped to 5, etc.  Because we used unique ordinals for each document being cached, those line up to single documents.

Then we just collect documents into the ordinals/centroids using your existing logic.

It's all a little complicated because this is somewhere in between breadth- and depth-first, which our agg framework doesn't really do out of the box right now.